BUG,DEP: Fixup quantile/percentile and rename interpolation->method

seberg opened the pull request:
This (unfortunately currently) does 3 things:

* It renames and deprecates the `interpolation` kwarg as `method`
* It fixes the documentation current plot to use the new methods (see below)
* It fixes two small issues with some discrete methods and tests them by expanding the "monoticity" test case

The main `method` docstring is modified to:
```
    method : str, optional
        This parameter specifies the method to use for estimating the
        percentile.  There are many different methods, some unique to NumPy.
        See the notes for explanation.  The options aligning with the R types
        and the H&F paper are:

        * (H&F 1): 'inverted_cdf'
        * (H&F 2): 'averaged_inverted_cdf'
        * (H&F 3): 'closest_observation'
        * (H&F 4): 'interpolated_inverted_cdf'
        * (H&F 5): 'hazen'
        * (H&F 6): 'weibull'
        * (H&F 7): 'linear'  (default)
        * (H&F 8): 'median_unbiased'
        * (H&F 9): 'normal_unbiased'

        Mainly for compatibility reasons, NumPy also supports the following
        options which appear to be unique to NumPy:

        * 'lower'
        * 'higher',
        * 'midpoint'
        * 'nearest'

        .. versionchanged:: 1.22.0
            This argument was previously called "interpolation" and only
            offered the "linear" default and last four options.
```

Here is the plot, note that it is smaller in the artifact: https://23201-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.percentile.html

![new-percentile-plot](https://user-images.githubusercontent.com/61977/140831414-b88a4954-fcd2-42fd-a79d-f92422d8b665.png)
(yeah, colors/linestyles are a bit annoying, some of the colors match, because they are somewhat related, I think)

Closes gh-20306.

seberg committed:
API,DEP: Rename percentile/quantile `interpolation=` to `method=`

seberg committed:
BUG: quantile discrete methods ended up using -1 as index sometimes

Also, the closest-observation did not correctly support multiple
quantiles calculated at the same time (broadcasting error).

seberg committed:
DOC: Fixup the percentile methods plot

seberg committed:
TST: Add deprecation testcase for quantile interpolation rename

seberg commented on numpy/lib/function_base.py:
```diff
@@ -4445,19 +4492,22 @@ def _lerp(a, b, t, out=None):
 def _get_gamma_mask(shape, default_value, conditioned_value, where):
     out = np.full(shape, default_value)
-    out[where] = conditioned_value
+    np.copyto(out, conditioned_value, where=where, casting="unsafe")
```
Normally `condition_value` is just a scalar, so this doesn't matter.  But in one of the branches, it is an array that matches the shape of `out`, so we have to use the `copyto` logic of copying both.  (To be honest, the casting is just for sports right now.)

seberg commented:
@bzah if you have some time, maybe you can also have a look?

BvB93 reviewed (commented):
Can you also rename the `interpolation` keyword in the [stub file](https://github.com/numpy/numpy/blob/main/numpy/lib/function_base.pyi)?
With the way things are currently set up you'll only have to update `np.percentile`.

seberg committed:
MAINT: Rename interpolation to method in percentile stubs

seberg committed:
DOC: Remove reference to paper from quantile `method` kwarg

Apparently, sphinx does not resolve references to footnotes from
parameter descriptions.

seberg committed:
DOC: Add ticks to quantile interpolation/method error

Co-authored-by: abel <aoun@cerfacs.fr>

bzah commented:
@seberg maybe it would be interesting to have a separate graph with only the discontinuous methods ?
Especially because someone (not sure of the ping) said he have some use of numpy's specific "lower" and "higher" methods.

For the record, the discontinuous methods are: "inverted_cdf", "averaged_inverted_cdf", "closest_observation", and "lower", "higher", "nearest" and "midpoint".
By keeping only the continuous methods from your graph, it would make it a bit easier to read as well.


The names are also quite confusing, without a graph one could potentially expect the same results from `averaged_inverted_cdf` and `midpoint` as well as from `nearest` and `closest_observation`.

seberg commented:
I guess I could do two plots, one with all the discrete and one with the non-discrete.  Although I wonder if it isn't also nice to mix.  Right, `closest_observation` seems the most confusing one.  Frankly, I wonder if it isn't worse than "midpoint", but I guess it probably has its purposes.  Should we consider naming it differently?

> Wolfram says: observation number closest to `q n`

(wikipedia is similar)  The tricky part is the `q n`, which I am not sure if it is obvious?

rossbar commented:
> I guess I could do two plots, one with all the discrete and one with the non-discrete.

I think there's merit in both approaches: having a single plot is nice because all of the information is collected in one place; however, as mentioned, the plot is quite busy. Two plots would have the advantage of being less busy, but the discontinuous distinction may not be particularly meaningful for the average user just trying to learn about the different methods.

FWIW I'd vote to add the plot as-is then evaluate alternatives in a follow-up PR to make comparison easier.

xor2k commented:
Just found another project place where they use the numpy-based notation, if this is of interest: https://www.tensorflow.org/probability/api_docs/python/tfp/stats/percentile

xor2k commented:
In the Numpy Zoom session today I've mentioned that we use 'lower' and 'upper' interpolation. We typically do so to compensate a bias and ensure that we obtain a value which has actually been measured (especially a value which is integral if all measurements are integral). If there are no NaNs, this can also be achieved by sorting along one axis and picking the target value, e.g. array entry 3 if there are 4 entries if one wants interpolation='upper' and q=0.5 ("slightly above median", especially for larger arrays). It would also be possible with interpolation='linear' and picking q=(3-1)/(4-1) = 2/3, but this could yield some numerical instability, with 2/3 being a good example.

Probably changing our codebase to np.sort would be more appropriate (and not too much of an effort) anyway. I was maybe a little bit too worried my favorite (interpolation) methods would go away in the future ;)

seberg committed:
DOC: Update percentile/quantile docs

Mainly fixes the method list slightly, tones down the warning a
bit and fixes the link to the paper (I did not realize that the
link failed to work due only because the reference was missing
from nanquantile/nanpercentile).

seberg committed:
DOC: Add release not for quantile `interpolation` rename

Also updates the old release note to include the info (and generally
tweaks it a bit)

seberg commented:
OK, updated, hopefully pretty much final, but especially release note bike shedding very welcome :):
1. tweaked the docstring, slightly less discouraging now (but still discouraging in the first paragraph which leads with "unique to NumPy").  Mention that some methods are discontinuous and all of the "bonus" ones are.
2. Tweaked the current release note.
3. Added a second deprecation release note.
4. Readded the reference to the paper to link from it, and use a direct enumeration.

seberg commented:
Btw. I won't worry about projects like tensorflow personally, they should be able to keep up.  We can't keep potentially serious flaws in what methods we offer stand in the way.
As for `pandas`, I pinged them, and they seem to fine with the deprecation.

charris commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -1230,3 +1230,23 @@ def test_deprecated_module(self):
+            [np.percentile, np.quantile, np.nanpercentile, np.nanquantile])
+    def test_both_passed(self, func):
+        with warnings.catch_warnings():
+            # catch the warning, but make sure it does not raise:
```
"does not" <- "does"?

seberg replied:
The warning is given before the error is thrown.  So we make sure the `DeprecationWarning` does _not_ raise, yes.  (In order to test the error that comes after the warning).

charris replied:
Maybe "catch the DeprecationWarning  so that it does not raise".

charris commented on numpy/lib/function_base.py:
```diff
@@ -61,13 +61,13 @@
 # (a.k.a 'g' or 'gamma')
 #
-# Each _QuantileInterpolation have two properties
+# Each _QuantileMethods has two properties
```
"Each method in"

seberg replied:
```suggestion
# Each method in _QuantileMethods has two properties
```

charris commented on numpy/lib/function_base.py:
```diff
@@ -3854,7 +3854,7 @@ def _median(a, axis=None, out=None, overwrite_input=False):
 
 def _percentile_dispatcher(a, q, axis=None, out=None, overwrite_input=None,
-                           interpolation=None, keepdims=None):
+                           method=None, keepdims=None, *, interpolation=None):
```
Hmm.  The change in order is probably safe. We should make the options keyword only at some point.

charris replied:
NVM, this is private, so no problem.

seberg replied:
This is the dispatch, so it is the same as the public signature.  This is something `__array_function__` implementers will have to catch up with.

I guess one annoyance is that it is hard to use avoid the warning if you are writing for both old and new NumPy versions.  That would be an argument to delay the `DeprecationWarning` for a bit...

charris replied:
Let's see how the first release goes.

charris commented on numpy/lib/function_base.py:
```diff
@@ -3864,8 +3864,10 @@ def percentile(a,
+               method="linear",
+               keepdims=False,
+               *,
+               interpolation=None):
```
Interesting approach,  replace "interpolation" by "method" when used as positional arguments and only worry if "interpolation" is used as a keyword.

seberg committed:
DOC: Fixups for interpolation rename comments from review

Co-authored-by: Charles Harris <charlesr.harris@gmail.com>

charris commented:
Thanks Sebastian.
